File: frontend/src/lib/api/approvals.ts
/**
 * Typed client for ``POST /api/sessions/{id}/approvals/{request_id}`` —
 * the approval-modal submit surface (``src/bearings/web/routes/approvals.py``
 * Slice A4).
 *
 * The endpoint returns 204 No Content on success. The caller receives
 * an :class:`ApiError` on 404 (no live broker), 409 (duplicate /
 * already-resolved request_id), or any other non-2xx.
 *
 * For standard tool approvals (``ApprovalModal``) only ``approved``
 * is sent. For ``AskUserQuestion`` approvals (``AskUserQuestionModal``)
 * ``answer`` carries the user's typed text; the backend threads it
 * back to the SDK callback as ``PermissionResultAllow.updated_input``
 * so the agent receives the answer as the tool result.
 */
import { sessionApprovalEndpoint } from "../config";
import { ApiError } from "./client";

const HTTP_OK_MIN = 200;
const HTTP_OK_MAX = 300;

/**
 * Resolve a pending ``can_use_tool`` approval request.
 *
 * @param sessionId  The session that owns the in-flight approval.
 * @param requestId  The ``request_id`` from the ``approval_request`` event.
 * @param approved   ``true`` to allow, ``false`` to deny.
 * @param answer     Optional text answer for ``AskUserQuestion`` approvals.
 */
export async function postApproval(
  sessionId: string,
  requestId: string,
  approved: boolean,
  answer?: string,
): Promise<void> {
  const body: { approved: boolean; answer?: string } = { approved };
  if (answer !== undefined) {
    body.answer = answer;
  }
  const response = await fetch(sessionApprovalEndpoint(sessionId, requestId), {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
    },
    body: JSON.stringify(body),
  });
  if (response.status < HTTP_OK_MIN || response.status >= HTTP_OK_MAX) {
    let errorBody: unknown = null;
    try {
      errorBody = await response.json();
    } catch {
      try {
        errorBody = await response.text();
      } catch {
        // ignore
      }
    }
    throw new ApiError(
      response.status,
      errorBody,
      `POST approval ${requestId} → ${response.status} ${response.statusText}`,
    );
  }
  // 204 No Content — nothing to parse.
}

File: src/bearings/web/routes/approvals.py
"""Approval-resolution route.

Per Slice A4 of ``~/.claude/plans/wiring-agent-loop.md``: when the
SDK opens ``can_use_tool`` mid-turn, the
:class:`bearings.agent.approval.ApprovalBroker` parks an
``asyncio.Future`` and emits an
:class:`bearings.agent.events.ApprovalRequest` so the conversation
pane opens its modal. The user clicks Allow or Deny; the frontend
POSTs the choice here. This route resolves the broker's future via
:meth:`ApprovalBroker.resolve` so the SDK callback returns and the
agent's turn proceeds.

Failure modes:

* ``404`` — no session matches the id, OR the session has no
  in-flight approval broker (e.g. the session has no live
  supervisor — the runner was reaped).
* ``409`` — the broker rejected the resolution (the request_id is
  unknown / already resolved). Per ``ApprovalBroker.resolve``'s
  defensive contract a duplicate REST + WS resolution surfaces as
  this 409 rather than silently double-approving.
"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, Request, status

from bearings.agent.approval import ApprovalBroker
from bearings.web.models.approvals import ApprovalResolution
from bearings.web.runner_factory import InProcessRunnerRegistry

router = APIRouter()


@router.post(
    "/api/sessions/{session_id}/approvals/{request_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def resolve_approval(
    session_id: str,
    request_id: str,
    payload: ApprovalResolution,
    request: Request,
) -> None:
    """Resolve a pending approval-modal request.

    The SDK callback waiting on the broker future receives
    ``payload.approved`` and translates it into a
    :class:`PermissionResultAllow` / :class:`PermissionResultDeny`
    SDK return value.
    """
    factory = getattr(request.app.state, "runner_factory", None)
    if not isinstance(factory, InProcessRunnerRegistry):
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="approval surface requires the in-process runner registry",
        )
    broker = factory.get_approval_broker(session_id)
    if broker is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"no approval broker for session {session_id!r}",
        )
    if not isinstance(broker, ApprovalBroker):  # pragma: no cover — type pin
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="approval broker has unexpected type",
        )
    resolved = await broker.resolve(request_id, approved=payload.approved, answer=payload.answer)
    if not resolved:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=(
                f"approval request {request_id!r} is unknown or already resolved "
                f"on session {session_id!r}"
            ),
        )


__all__ = ["router"]
